server/src/main/java/org/apache/lucene/search/ConcurrentTopScoreDocCollector.java

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.apache.lucene.search;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.MaxScoreAccumulator.DocAndScore;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * This is a modified version of {@link TopScoreDocCollector} that doesn't keep any leaf-specific state in
+ * the collector and therefore its leaf collectors can be used concurrently.
```

Comment:
We would need to create it for each segment. Also we don't need a full concurrency, as it will be executed on a single thread, but we are jumping between leaves before finishing individual leaves, so we cannot share any state between leaves on the collector level.

Reply:
I think the idea of `CollectorManager` is that it's an abstraction around *how* iteration works. When you put it like that it feels appropriate.